sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,388 @@
+
+import LibSodium.Bindings.Random (randombytesBuf)
+import LibSodium.Bindings.Secretbox (cryptoSecretboxEasy, cryptoSecretboxKeyBytes, cryptoSecretboxKeygen, cryptoSecretboxMACBytes, cryptoSecretboxNonceBytes, cryptoSecretboxOpenEasy)
+import Sel.Internal
+
+-- $introduction
+-- "Authenticated Encryption" uses a secret key along with a single-use number
+-- called a "nonce" to encrypt a message.
+-- The resulting hash is accompanied by an authentication tag.
+--
+-- Encryption is done with the XSalsa20 stream cipher and authentication is done with the
+-- Poly1305 MAC hash.
+
+-- $usage
+--
+-- > import qualified Sel.SecretKey.AuthenticatedEncryption as AuthenticatedEncryption
+-- >
+-- > main = do
+-- >   -- We get the secretKey from the other party or with 'newSecretKey'.
+-- >   -- We get the nonce from the other party with the message, or with 'encrypt' and our own message.
+-- >   -- Do not reuse a nonce with the same secret key!
+-- >   (nonce, encryptedMessage) <- AuthenticatedEncryption.encrypt "hello hello" secretKey
+-- >   let result = AuthenticatedEncryption.decrypt encryptedMessage secretKey nonce
+-- >   print result
+-- >   -- "Just \"hello hello\""
+
+-- | A secret key of size 'cryptoSecretboxKeyBytes'.
+--
+-- @since 0.0.1.0
+newtype SecretKey = SecretKey (ForeignPtr CUChar)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq SecretKey where
+  (SecretKey hk1) == (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord SecretKey where
+  compare (SecretKey hk1) (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Display SecretKey where
+  displayBuilder = Builder.fromText . secretKeyToHexText
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Show SecretKey where
+  show = BS.unpackChars . secretKeyToHexByteString
+
+-- | Generate a new random secret key.
+--
+-- @since 0.0.1.0
+newSecretKey :: IO SecretKey
+newSecretKey = do
+  fPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxKeyBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    cryptoSecretboxKeygen ptr
+  pure $ SecretKey fPtr
+
+-- | Create a 'SecretKey' from a 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- @since 0.0.1.0
+secretKeyFromByteString :: StrictByteString -> SecretKey
+secretKeyFromByteString bytestring = unsafeDupablePerformIO $
+  BS.unsafeUseAsCStringLen bytestring $ \(outsideSecretKeyPtr, _) -> do
+    secretKeyForeignPtr <- BS.mallocByteString @CChar (fromIntegral cryptoSecretboxKeyBytes)
+    Foreign.withForeignPtr secretKeyForeignPtr $ \secretKeyPtr ->
+      Foreign.copyArray outsideSecretKeyPtr secretKeyPtr (fromIntegral cryptoSecretboxKeyBytes)
+    pure $ SecretKey (Foreign.castForeignPtr @CChar @CUChar secretKeyForeignPtr)
```

Comment:
Will this accept _absolutely any_ `StrictByteString`? I assume there's a length requirement.